# Incidence Matrixes

In [2]:
sample_bbc_news_sentences = [
    "China confirms Interpol chief detained",
    "Turkish officials believe the Washington Post writer was killed in the Saudi consulate in Istanbul.",
    "US wedding limousine crash kills 20",
    "Bulgarian journalist killed in park",
    "Kanye West deletes social media profiles",
    "Brazilians vote in polarised election",
    "Bull kills woman at French festival",
    "Indonesia to wrap up tsunami search",
    "Tina Turner reveals wedding night ordeal",
    "Victory for Trump in Supreme Court battle",
    "Clashes at German far-right rock concert",
    "The Walking Dead, actor dies aged 76",
    "Jogger in Netherlands finds lion cub",
    "Monkey takes the wheel of Indian bus"
]

In [3]:
#basic tokenization
from nltk.tokenize import TweetTokenizer

tokenizer = TweetTokenizer()
sample_bbc_news_sentences_tokenized = [tokenizer.tokenize(sentence) 
                            for sentence in sample_bbc_news_sentences]
temp=[]
for sentence in sample_bbc_news_sentences:
    temp.append(tokenizer.tokenize(sentence))

sample_bbc_news_sentences_tokenized[0]

['China', 'confirms', 'Interpol', 'chief', 'detained']

In [4]:
sample_bbc_news_sentences_tokenized_lower = [[_t.lower() 
                                              for _t in _s] 
                for _s in sample_bbc_news_sentences_tokenized]

sample_bbc_news_sentences_tokenized_lower[0]


['china', 'confirms', 'interpol', 'chief', 'detained']

In [6]:
#get all unique tokens
# Step 1: Flatten (sum)
# Step 2: Put in SET

# [[token1,token2,...],[...],[...],...]
unique_tokens = set(sum(sample_bbc_news_sentences_tokenized_lower, []))
# [token1, token2,...,token_K+1]
print(list(unique_tokens)[:5])
print(len(unique_tokens))

['interpol', 'istanbul', 'rock', 'turner', 'victory']
83


In [7]:
# create incidence matrix (term-document frequency)
import numpy as np
incidence_matrix = np.array([[sent.count(token)
                      for sent in sample_bbc_news_sentences_tokenized_lower] 
                for token in unique_tokens])
a =[]
b = []
for sent in sample_bbc_news_sentences_tokenized_lower:
    for token in unique_tokens:
        a.append(sent.count(token))
    b.append(a)
    a = []
print(incidence_matrix)
#incidence_matrix.shape

[[1 0 0 ... 0 0 0]
 [0 1 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [8]:
incidence_matrix.shape

(83, 14)

![Inverted Index](img/inverted-index.png)

# Dataset 
https://archive.ics.uci.edu/ml/datasets/Twenty+Newsgroups 

# You will now have to construct the Inverted Index - only the dictionary part (term and #docs)

In [9]:
from nltk.tokenize import sent_tokenize
from collections import defaultdict, Counter
from string import punctuation
import os
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/petar.kolev/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [16]:
document = "ASdlkasjlkdja lkasjdlka lkdjas dlka jdslkjsa dla ljalkd lkas lalkjdslkajs ld!"


def preprocess_document(content):
    """
    Returns a list of tokens for a document's content. 
    Tokens should not contain punctuation and should be lower-cased.
    """
    tokens = content.split()

    tokens = [''.join(ch for ch in token if ch not in punctuation).lower()
                for token in tokens if ''.join(ch for ch in token if ch not in punctuation)]

    return tokens


def prepare_dataset(documents_dir):
    """
    Returns list of documents in the documents_dir, where each document is a list of its tokens. 
    
    """
    dataset = []
    for filename in os.listdir(documents_dir):
        file_path = os.path.join(documents_dir, filename)
        if os.path.isfile(file_path):
            with open(file_path) as f:
                content = f.read()
                tokens = preprocess_document(content)
                dataset.append(tokens)

    return dataset

In [15]:
preprocess_document("US wedding limousine crash kills 20. Kanye West, deletes social media profiles!")

['us',
 'wedding',
 'limousine',
 'crash',
 'kills',
 '20',
 'kanye',
 'west',
 'deletes',
 'social',
 'media',
 'profiles']

In [ ]:
prepare_dataset('./data/mini_newsgroups/sci.electronics')

[['newsgroups',
  'scielectronics',
  'path',
  'cantaloupesrvcscmueducrabapplesrvcscmuedufs7ececmuedueuropaenggtefsdcomgatechukmabillq',
  'from',
  'billqmsukyedu',
  'billy',
  'quinn',
  'subject',
  'suggestions',
  'on',
  'audio',
  'relays',
  'messageid',
  'c5qsbfiekmsukyedu',
  'organization',
  'university',
  'of',
  'kentucky',
  'dept',
  'of',
  'math',
  'sciences',
  'date',
  'mon',
  '19',
  'apr',
  '1993',
  '175225',
  'gmt',
  'lines',
  '22',
  'i',
  'built',
  'a',
  'little',
  'project',
  'using',
  'the',
  'radio',
  'shack',
  '5vdc',
  'relays',
  'to',
  'switch',
  'audio',
  'i',
  'got',
  'pretty',
  'bad',
  'clicks',
  'when',
  'the',
  'thing',
  'switched',
  'i',
  'was',
  'doing',
  'most',
  'of',
  'the',
  'common',
  'things',
  'one',
  'is',
  'supposed',
  'to',
  'do',
  'when',
  'using',
  'relays',
  'and',
  'nothing',
  'seemed',
  'to',
  'get',
  'rid',
  'of',
  'the',
  'clicks',
  'my',
  'question',
  'is',
  'is',
  'th

In [20]:
def document_frequency(tokenized_documents):
    """
    Returns a dictionary {token:number of documents containing the token}
    """
    df = defaultdict(int)
    for doc in tokenized_documents:
        unique_tokens = set(doc)
        for token in unique_tokens:
            df[token] += 1

    return dict(df)

In [21]:
selected_category = 'data/mini_newsgroups/sci.crypt/'
print(selected_category)
tokenized_dataset = prepare_dataset(selected_category)
print("Sample tokenized document:")
print(tokenized_dataset[0][:10])

df = document_frequency(tokenized_dataset)
print(df)

data/mini_newsgroups/sci.crypt/
Sample tokenized document:
['newsgroups', 'scicrypt', 'path', 'cantaloupesrvcscmuedurochesterudelbogussuranethowlandrestonansnettornnottbnrgatebnrcoukuknetpavocsicamacukgtf1000', 'from', 'gtf1000cuscamacuk', 'gt', 'falk', 'subject', 'altsecuritypgp']
{'england': 1, 'logging': 1, 'a': 81, 'g': 7, 'falk': 1, 'session': 7, '102016': 1, 'thanks': 5, 'version': 9, 'encrypted': 15, 'network': 8, 'after': 14, 'gw7z': 1, 'publickey': 5, 'apr': 100, 'the': 97, 'ajiif6tsmymqbwmvholm7bub4jpybqanpkmz8tdd4tyuinsx68cvg': 1, 'while': 19, 'nntppostinghost': 37, 'bootescuscamacuk': 1, '25': 11, 'me': 28, 'my': 35, 'report': 5, 'path': 100, '22': 23, 'messages': 11, 'gmt': 84, 'signature': 3, 'pgp': 12, 'does': 31, 'purportedly': 1, 'lines': 99, 'sent': 6, 'remotely': 1, 'newsgroups': 100, 'end': 8, 'or': 56, '220493': 1, 'password': 2, 'in': 84, '1993apr221020164080infodevcamacuk': 1, 'know': 34, 'hence': 4, 'messageid': 100, 'unauthorised': 2, 'nobody': 4, '1993': 75, '

In [22]:

print("Most common words:")
print(Counter(df).most_common(10))
print("Least common words:")
print(Counter(df).most_common()[-10:])

Most common words:
[('apr', 100), ('path', 100), ('newsgroups', 100), ('messageid', 100), ('date', 100), ('subject', 100), ('from', 100), ('lines', 99), ('the', 97), ('organization', 94)]
Least common words:
[('invisible', 1), ('myktotronx', 1), ('pincompatible', 1), ('pink', 1), ('speculate', 1), ('hallucinating', 1), ('intercon', 1), ('cipher2', 1), ('altsecuritypgp2756', 1), ('refers', 1)]


## __Questions__: 
- Which words have highest and which lowest Total freq?
- What are some examples of inverted indexes we have seen in life?
- Why do we use inverted indexes?

![Sort-Based-Index](img/sort-based-method.png)

In [23]:
!pip install nltk

# Preprocessing:

In [24]:
from nltk.tokenize import sent_tokenize, TweetTokenizer
import nltk
nltk.download('punkt_tab')
from string import punctuation
tokenizer = TweetTokenizer()

def preprocess_document(content):
    """
    Returns a list of tokens for a document's content. 
    Tokens should not contain punctuation and should be lower-cased.
    """
    sentences = sent_tokenize(content)
    tokens = []
    for _sent in sentences:
        sent_tokens = tokenizer.tokenize(_sent)
        sent_tokens = [_tok.lower() for _tok in sent_tokens if _tok not in punctuation]
        tokens += sent_tokens
    return tokens

[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/petar.kolev/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [25]:
preprocess_document(open('data/mini_newsgroups/rec.autos/101629').read())[:10]

['path',
 'cantaloupe.srv.cs.cmu.edu',
 'crabapple.srv.cs.cmu.edu',
 'fs7.ece.cmu.edu',
 'europa.eng.gtefsd.com',
 'howland.reston.ans.net',
 'wupost',
 'uunet',
 'caen',
 'rphroy']

# Extract Pairs of (token, document_id) tuples 
These will eventually end up sorted by document_id.

In [26]:
from os import scandir # can be used for easier iteration of documents in a folder
# can check is_file() on the objects returned by scan_dir 
# contain whole document path, so no need to join with the directory

import os

def get_token_doc_id_pairs(category_dir):
    """
    Iteratively goes through the documents in the category_dir and constructs/returns:
    1. A list of (token, doc_id) tuples
    2. A dictionary of doc_id:doc_name
    """
    token_doc_pairs = []
    doc_id_to_name = {}

    doc_id = 0
    for entry in scandir(category_dir):
        if entry.is_file():
            doc_name = entry.name
            doc_id_to_name[doc_id] = doc_name

            # Read and preprocess the file
            with open(entry.path, encoding='utf-8') as f:
                content = f.read()
                tokens = preprocess_document(content)

            # Add (token, doc_id) pairs
            token_doc_pairs.extend((token, doc_id) for token in tokens)

            doc_id += 1

    return token_doc_pairs, doc_id_to_name

In [27]:
token_docid, doc_ids = get_token_doc_id_pairs('data/mini_newsgroups/rec.autos/')

In [28]:
print(doc_ids[2])

101577


In [29]:
token_docid[:10]

[('newsgroups', 0),
 ('rec.autos', 0),
 ('path', 0),
 ('cantaloupe.srv.cs.cmu.edu', 0),
 ('magnesium.club.cc.cmu.edu', 0),
 ('news.sei.cmu.edu', 0),
 ('fs7.ece.cmu.edu', 0),
 ('europa.eng.gtefsd.com', 0),
 ('howland.reston.ans.net', 0),
 ('ux1.cso.uiuc.edu', 0)]

__Example output:__ <br>
token_docid, doc_ids = get_token_doc_id_pairs('data/mini_newsgroups/rec.autos/')<br>
print(doc_ids[2])<br>
token_docid[:10]<br>

> DirEntry '101577' <br>
>[('newsgroups', 0),
> ('rec.autos', 0),
> ('path', 0),
> ('cantaloupe.srv.cs.cmu.edu', 0),
> ('magnesium.club.cc.cmu.edu', 0),
> ('news.sei.cmu.edu', 0),
> ('fs7.ece.cmu.edu', 0),
> ('europa.eng.gtefsd.com', 0),
> ('howland.reston.ans.net', 0),
> ('ux1.cso.uiuc.edu', 0)]

# Sort by token

In [30]:
from operator import itemgetter
from collections import defaultdict
sorted_token_docid = sorted(token_docid, key=itemgetter(0))
sorted_token_docid[-10:]

[('zaphod.mps.ohio-state.edu', 97),
 ('zaphod.mps.ohio-state.edu', 98),
 ('zaphod.mps.ohio-state.edu', 99),
 ('zauberer', 47),
 ('zeolite', 49),
 ('zip', 77),
 ('zip', 77),
 ('zx', 86),
 ('zx-r', 57),
 ("|'8", 70)]

# Merge token occurrences in a single document -> (token, doc_id, term_freq) tuples

In [31]:
def merge_token_in_doc(sorted_token_docid):
    """
    Returns a list of (token, doc_id, term_freq) tuples from a sorted list of (token, doc_id) list, 
    where if a token appears n times in a doc_id, we merge it in a tuple (toke, doc_id, n).
    """
    if not sorted_token_docid:
        return []

    merged = []
    current_token, current_doc, count = sorted_token_docid[0][0], sorted_token_docid[0][1], 1

    for token, doc_id in sorted_token_docid[1:]:
        if token == current_token and doc_id == current_doc:
            count += 1
        else:
            merged.append((current_token, current_doc, count))
            current_token, current_doc, count = token, doc_id, 1

    # append the last group
    merged.append((current_token, current_doc, count))
    return merged

In [32]:
merged_token_in_docs = merge_token_in_doc(sorted_token_docid)
merged_token_in_docs[-10:]

[('zaphod.mps.ohio-state.edu', 96, 1),
 ('zaphod.mps.ohio-state.edu', 97, 1),
 ('zaphod.mps.ohio-state.edu', 98, 1),
 ('zaphod.mps.ohio-state.edu', 99, 1),
 ('zauberer', 47, 1),
 ('zeolite', 49, 1),
 ('zip', 77, 2),
 ('zx', 86, 1),
 ('zx-r', 57, 1),
 ("|'8", 70, 1)]

__Example output:__ <br>

merge_token_in_doc = merge_token_in_doc(sorted_token_docid) <br>
merged_tokens_in_doc[-10:] <br>

>[('zaphod.mps.ohio-state.edu', 96, 1),
 ('zaphod.mps.ohio-state.edu', 97, 1),
 ('zaphod.mps.ohio-state.edu', 98, 1),
 ('zaphod.mps.ohio-state.edu', 99, 1),
 ('zauberer', 47, 1),
 ('zeolite', 49, 1),
 ('zip', 77, 2),
 ('zx', 86, 1),
 ('zx-r', 57, 1),
 ("|'8", 70, 1)]

# Split into Dictionary and Postings (usually linked lists for each word)

In [33]:
from collections import defaultdict
dictionary = defaultdict(lambda: (0, 0)) # term : doc_freq, tot freq
postings = defaultdict(lambda: []) # term: doc_ids, doc_freq

for token, doc_id, doc_freq in merged_token_in_docs:
    dictionary[token] = (dictionary[token][0]+1, dictionary[token][1]+doc_freq)

# usually implemented as linked lists
for token, doc_id, doc_freq in merged_token_in_docs:
    postings[token].append((doc_id, doc_freq)) 

In [34]:
dictionary['zip'], dictionary['zaphod.mps.ohio-state.edu']

((1, 2), (51, 52))

In [35]:
postings['zip'], postings['zaphod.mps.ohio-state.edu'][:3]

([(77, 2)], [(1, 1), (3, 1), (4, 1)])

# Boolean Queries in the index

## AND query: we want to find documents which contain both 'living' and 'dead' in them.

We use a merging algorithm for conjunction queries, which simultaneously traverses the postings of the given words.

- Living: 1 -> 2 -> 5 -> 17 -> 30 -> 31 -> 44 -> 45 -> 47
- Dead: 5 -> 17 -> 44
- Intersection: 5 -> 17 -> 44

It takes __linear time__ over the number of documents the two words appear in. <br/>
It is important to have postings for each word __sorted by document id__.

In [36]:
def and_query(postings, word1, word2):
    """
    merging postings lists of two words
    """
    p1 = postings.get(word1, [])
    p2 = postings.get(word2, [])
    i = j = 0
    out = []

    while i < len(p1) and j < len(p2):
        if p1[i] == p2[j]:
            out.append(p1[i])
            i += 1
            j += 1
        elif p1[i] < p2[j]:
            i += 1
        else:
            j += 1
    return out

In [37]:
doc_id = and_query(postings, 'living', 'dead')
doc_id

[(42, 1)]

In [38]:
postings['living']

[(7, 1), (42, 1), (64, 1)]

In [39]:
postings['dead']

[(12, 1), (42, 1), (82, 1)]

__Example Output:__ <br>

doc_id = and_query(postings, 'living', 'dead') <br>
doc_ids[doc_id[0]] <br>

> DirEntry '102983'

## Questions:

- How about if we want to find a document containing N words? 
- What will be the execution time for queries with NOT/ OR and different combinations?
- What are the downsides of boolean queries?

## Advanced query features
* We can optimize with : processing in order of increasing freq; skip pointers
* Proximity
* Zones
* Phrase queries (bi-word indexes, positional indexes)